liquidity-rewards/contracts/UniswapPoolRewards.sol

```diff
@@ -1,5 +1,193 @@
+
+        super.stake(amount);
+        periodTotalStaked = periodTotalStaked.add(amount);
+        emit Staked(msg.sender, rewardPeriod, amount);
+    }
+
+    function withdraw(uint256 amount) public override{
+        require(amount > 0, "Cannot withdraw 0");
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updateAccountReward(msg.sender);
+
+        super.withdraw(amount);
+        periodTotalStaked = periodTotalStaked.sub(amount);
+        emit Withdrawn(msg.sender, rewardPeriod, amount);
+    }
+
+    // Shortcut to be able to unstake tokens and claim rewards in one transaction
+    function withdrawAndClaim() external {
+        withdraw(balanceOf(msg.sender));
+        claimReward();
+    }
+
+    function claimReward() public {
+        require(address(lpToken) != address(0), "Liquidity Pool Token has not been set yet");
+
+        _updatePeriodFinish();
+        _updateAccountReward(msg.sender);
+
+        uint256 reward = earned(msg.sender);
+
+        require(reward > 0, "Nothing to claim");
+
+        rewards[msg.sender] = 0;
+        totalRewardPaid = totalRewardPaid.add(reward);
+        periodRewardPaid = periodRewardPaid.add(reward);
+        elimuToken.transfer(msg.sender, reward);
+        emit RewardPaid(msg.sender, rewardPeriod, reward);
+    }
+
+
+    function _notifyRewardAmount(uint256 _reward, uint256 _duration) internal {
+        assert(_reward > 0);
+        elimuToken.transferFrom(msg.sender, address(this), _reward);
+        
+        uint256 totalReward = elimuToken.balanceOf(address(this));
+        _updateReward();
+
+        rewardRate = totalReward.div(_duration);
+
+        lastUpdateTime = block.timestamp;
+        periodFinish = block.timestamp.add(_duration);
+    }
+
+    // Adjusts end time for the program after periods of zero total supply
+    function _updatePeriodFinish() internal {
+        if (totalSupply() == 0) {
+            assert(periodFinish > 0);
+            /*
+             * If the finish period has been reached (but there are remaining rewards due to zero stake),
+             * to get the new finish date we must add to the current timestamp the difference between
+             * the original finish time and the last update, i.e.:
+             *
+             * periodFinish = block.timestamp.add(periodFinish.sub(lastUpdateTime));
+             *
+             * If we have not reached the end yet, we must extend it by adding to it the difference between
+             * the current timestamp and the last update (the period where the supply has been empty), i.e.:
+             *
+             * periodFinish = periodFinish.add(block.timestamp.sub(lastUpdateTime));
+             *
+             * Both formulas are equivalent.
+             */
+            periodFinish = periodFinish.add(block.timestamp.sub(lastUpdateTime));
+        }
+    }
+
+    function _updateReward() internal {
+        rewardPerTokenStored = rewardPerToken();
```

Comment:
Should this variable be called `rewardPerTokenDeposited`, or am I misunderstanding the use of the word "Stored" here?

Reply:
Ok, we could change it to that too!